Superdense coding kata finish (#1479)
microsoft/qsharp: katas/content/superdense_coding/alice_sends_message/solution.md

Recall that we learnt how to prepare all Bell states in "Preparing Quantum States" kata. This is slightly advanced version of that task demonstrating the "power of entanglement", where operations applied to only one qubit affect the state of the system in interesting way.

Let's use the below Bell state we prepared in the previous task:

$$\frac{1}{\sqrt{2}} \big(\ket{00} + \ket{11} \big)$$

> We can transform it according to the value of bits:

Reviewer: ```suggestion
We can transform it into every other Bell state according to the value of `message`:
```

> - `(0, 0)`: $\ket{\Phi^{+}} = \frac{1}{\sqrt{2}} (\ket{00} + \ket{11})$

Reviewer: These two lists are rendered as one 8-point list, which looks a bit odd. I would merge the pairs of bullet points that describe the same scenario into one point each, so that each value of message is described and handled right away

Author: Nice catch. Updated solution.

- `(0, 1)`: $\ket{\Psi^{+}} = \frac{1}{\sqrt{2}} (\ket{01} + \ket{10})$
- `(1, 0)`: $\ket{\Phi^{-}} = \frac{1}{\sqrt{2}} (\ket{00} - \ket{11})$
- `(1, 1)`: $\ket{\Psi^{-}} = \frac{1}{\sqrt{2}} (\ket{01} - \ket{10})$

- If `bits == (0, 0)`, we do nothing - the prepared state is already $\ket{\Phi^{+}}$.

- If `bits ==  (0, 1)`, we need to change the second qubit in both $\ket{00}$ and $\ket{11}$ terms. Observe that applying an $X$ gate to Alice's qubit does exactly that:

>   $$X \otimes I (\ket{\Phi^{+}}) = \frac{1}{\sqrt{2}} (\ket{10} + \ket{01})$$

Reviewer: Here and in the next formula we need brackets around (X \otimes I) rather than around the \ket for Bell state (just like you did in the last formula) - \ket is already a single entity, but getting a two-qubit gate out of the tensor product needs emphasizing

Author: Done

- If `bits == (1, 0)`, we need to add a relative phase of $-1$ to the $\ket{11}$ term. Observe that applying $Z$ gate to Alice's qubit does exactly that:
  $$Z \otimes I (\ket{\Phi^{+}}) = \frac{1}{\sqrt{2}} (\ket{00} - \ket{11})$$

- If `bits = (1, 1)`, we use the same logic to realize that we need to apply both the $Z$ and $X$ corrections to get $\ket{\Psi^{-}}$ state.

>   $$ (Z \otimes I).(X \otimes I) (\ket{\Psi^{+}}) = (Z \otimes I) \frac{1}{\sqrt{2}} (\ket{10} + \ket{01}) = \frac{1}{\sqrt{2}} (-\ket{10} + \ket{01}) = \frac{1}{\sqrt{2}} (\ket{01} - \ket{10}) $$

Reviewer: ```suggestion
  $$ (Z \otimes I) \cdot (X \otimes I) \ket{\Psi^{+}} = (Z \otimes I) \frac{1}{\sqrt{2}} (\ket{10} + \ket{01}) = \frac{1}{\sqrt{2}} (-\ket{10} + \ket{01}) = \frac{1}{\sqrt{2}} (\ket{01} - \ket{10}) $$
```

The final sequence of steps is as follows:

> 1. Apply the $X$ gate to the alice's qubit if `bit2 == 1`.
> 2. Apply the $Z$ gate to the alice's qubit if `bit1 == 1`.

Reviewer: ```suggestion
1. Apply the $X$ gate to Alice's qubit if `bit2 == 1`.
2. Apply the $Z$ gate to Alice's qubit if `bit1 == 1`.
```

Recall that we learnt how to prepare all Bell states in "Preparing Quantum States" kata. This is slightly advanced version of that task demonstrating how operations applied to one qubit of a Bell state allow us to transform it into any other Bell state.

Superdense coding protocol uses the below Bell state we prepared in the previous task:

$$\frac{1}{\sqrt{2}} \big(\ket{00} + \ket{11} \big)$$

We can transform it into every other Bell state according to the value of `message`:

- `(0, 0)`: $\ket{\Phi^{+}} = \frac{1}{\sqrt{2}} (\ket{00} + \ket{11})$
- `(0, 1)`: $\ket{\Psi^{+}} = \frac{1}{\sqrt{2}} (\ket{01} + \ket{10})$
- `(1, 0)`: $\ket{\Phi^{-}} = \frac{1}{\sqrt{2}} (\ket{00} - \ket{11})$
- `(1, 1)`: $\ket{\Psi^{-}} = \frac{1}{\sqrt{2}} (\ket{01} - \ket{10})$

> ### Detailed Explaination

Reviewer: We use subheaders as headers, when there's a different solution or a different topic raised. Here, the goal of the solution is already to provide a detailed explanation, so I'd rephrase
```suggestion
Here is how we can perform this transformation:
```
